File: Auto/car_app_library/showcase/src/main/java/com/google/android/libraries/car/app/samples/showcase/StartScreen.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.libraries.car.app.samples.showcase;

import androidx.annotation.NonNull;
import androidx.core.graphics.drawable.IconCompat;
import com.google.android.libraries.car.app.CarContext;
import com.google.android.libraries.car.app.Screen;
import com.google.android.libraries.car.app.model.Action;
import com.google.android.libraries.car.app.model.CarIcon;
import com.google.android.libraries.car.app.model.ItemList;
import com.google.android.libraries.car.app.model.ListTemplate;
import com.google.android.libraries.car.app.model.Row;
import com.google.android.libraries.car.app.model.Template;
import com.google.android.libraries.car.app.samples.showcase.misc.MiscDemoScreen;
import com.google.android.libraries.car.app.samples.showcase.navigation.NavigationDemosScreen;
import com.google.android.libraries.car.app.samples.showcase.templates.MiscTemplateDemosScreen;
import com.google.android.libraries.car.app.samples.showcase.textandicons.TextAndIconsDemosScreen;

/** The starting screen of the app. */
public final class StartScreen extends Screen {
  public StartScreen(CarContext carContext) {
    super(carContext);
  }

  @NonNull
  @Override
  public Template getTemplate() {
    ItemList.Builder listBuilder = ItemList.builder();

    listBuilder.addItem(
        Row.builder()
            .setTitle("Selectable Lists Demo")
            .setOnClickListener(
                () -> getScreenManager().push(new SelectableListsDemoScreen(getCarContext())))
            .build());

    listBuilder.addItem(
        Row.builder()
            .setTitle("Task Restriction Demo")
            .setOnClickListener(
                () -> getScreenManager().push(new TaskRestrictionDemoScreen(1, getCarContext())))
            .build());

    listBuilder.addItem(
        Row.builder()
            .setImage(
                CarIcon.of(
                    IconCompat.createWithResource(getCarContext(), R.drawable.ic_map_white_48dp)))
            .setTitle("Navigation Demos")
            .setOnClickListener(
                () -> {
                  getScreenManager().push(new NavigationDemosScreen(getCarContext()));
                })
            .setIsBrowsable(true)
            .build());

    listBuilder.addItem(
        Row.builder()
            .setTitle("Misc Templates Demos")
            .setOnClickListener(
                () -> getScreenManager().push(new MiscTemplateDemosScreen(getCarContext())))
            .setIsBrowsable(true)
            .build());

    listBuilder.addItem(
        Row.builder()
            .setTitle("Text and Icons Demos")
            .setOnClickListener(
                () -> getScreenManager().push(new TextAndIconsDemosScreen(getCarContext())))
            .setIsBrowsable(true)
            .build());

    listBuilder.addItem(
        Row.builder()
            .setTitle("Misc Demos")
            .setOnClickListener(() -> getScreenManager().push(new MiscDemoScreen(getCarContext())))
            .setIsBrowsable(true)
            .build());

    return ListTemplate.builder()
        .setSingleList(listBuilder.build())
        .setTitle("Showcase Demos")
        .setHeaderAction(Action.APP_ICON)
        .build();
  }
}

File: car_app_library/showcase/shared/src/main/java/androidx/car/app/sample/showcase/common/StartScreen.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.car.app.sample.showcase.common

import androidx.car.app.CarContext
import androidx.car.app.Screen
import androidx.car.app.constraints.ConstraintManager
import androidx.car.app.model.Action
import androidx.car.app.model.CarIcon
import androidx.car.app.model.ItemList
import androidx.car.app.model.ListTemplate
import androidx.car.app.model.Row
import androidx.car.app.model.Template
import androidx.car.app.sample.showcase.common.misc.MiscDemoScreen
import androidx.car.app.sample.showcase.common.navigation.NavigationDemosScreen
import androidx.car.app.sample.showcase.common.templates.MiscTemplateDemosScreen
import androidx.car.app.sample.showcase.common.textandicons.TextAndIconsDemosScreen
import androidx.car.app.versioning.CarAppApiLevels
import androidx.core.graphics.drawable.IconCompat

/** The starting screen of the app.  */
class StartScreen(carContext: CarContext, private val mShowcaseSession: ShowcaseSession) :
    Screen(carContext) {
    override fun onGetTemplate(): Template {
        val listBuilder = ItemList.Builder()
        listBuilder.addItem(
            Row.Builder()
                .setTitle(carContext.getString(R.string.selectable_lists_demo_title))
                .setOnClickListener {
                    screenManager
                        .push(
                            SelectableListsDemoScreen(
                                carContext
                            )
                        )
                }
                .build())
        listBuilder.addItem(
            Row.Builder()
                .setTitle(carContext.getString(R.string.task_restriction_demo_title))
                .setOnClickListener {
                    screenManager
                        .push(
                            TaskRestrictionDemoScreen(
                                1, carContext
                            )
                        )
                }
                .build())
        listBuilder.addItem(
            Row.Builder()
                .setImage(
                    CarIcon.Builder(
                        IconCompat.createWithResource(
                            carContext,
                            R.drawable.ic_map_white_48dp
                        )
                    )
                        .build(),
                    Row.IMAGE_TYPE_ICON
                )
                .setTitle(carContext.getString(R.string.nav_demos_title))
                .setOnClickListener {
                    screenManager
                        .push(NavigationDemosScreen(carContext))
                }
                .setBrowsable(true)
                .build())
        var listLimit = 6
        // Adjust the item limit according to the car constrains.
        if (carContext.carAppApiLevel > CarAppApiLevels.LEVEL_1) {
            listLimit = carContext.getCarService(ConstraintManager::class.java).getContentLimit(
                ConstraintManager.CONTENT_LIMIT_TYPE_LIST
            )
        }
        val miscTemplateDemoScreenItemLimit = listLimit
        listBuilder.addItem(
            Row.Builder()
                .setTitle(carContext.getString(R.string.misc_templates_demos_title))
                .setOnClickListener {
                    screenManager
                        .push(
                            MiscTemplateDemosScreen(
                                carContext,
                                0,
                                miscTemplateDemoScreenItemLimit
                            )
                        )
                }
                .setBrowsable(true)
                .build())
        listBuilder.addItem(
            Row.Builder()
                .setTitle(carContext.getString(R.string.text_icons_demo_title))
                .setOnClickListener {
                    screenManager
                        .push(TextAndIconsDemosScreen(carContext))
                }
                .setBrowsable(true)
                .build())
        listBuilder.addItem(
            Row.Builder()
                .setTitle(carContext.getString(R.string.misc_demo_title))
                .setOnClickListener {
                    screenManager
                        .push(
                            MiscDemoScreen(
                                carContext,
                                mShowcaseSession
                            )
                        )
                }
                .setBrowsable(true)
                .build())
        return ListTemplate.Builder()
            .setSingleList(listBuilder.build())
            .setTitle(
                carContext.getString(R.string.showcase_demos_title) + " ("
                        + carContext.getString(
                    R.string.cal_api_level_prefix,
                    carContext.carAppApiLevel
                ) + ")"
            )
            .setHeaderAction(Action.APP_ICON)
            .build()
    }
}